rfc/0020-policy-group.md

```diff
@@ -0,0 +1,284 @@
+
+After the validation step, precompiled policies and group settings will be added to the `EvaluationEnvironment`.
+When a validation request is received, if the policy is a group policy, the `EvaluationEnvironment` performs the following steps:
+
+- create a new Rhai [raw Engine](https://rhai.rs/book/engine/raw.html?highlight=raw%20eng#raw-engine)
+- bind native functions to the engine that validate the policy
+- call the engine eval method with the expression
+- return the result
+
+Creating a new engine for each request is desirable because it avoids contention and locking of the Engine instance.
+This pattern is described in the Rhai book, under the [One Engine Instance Per Call](https://rhai.rs/book/patterns/parallel.html?highlight=one%20engine%20per#one-engine-instance-per-call) section.
+
+To improve performance, we could consider precompiling the AST of the expression when building the `EvaluationEnvironment`.
+Also, we could consider creating a [custom package](https://rhai.rs/book/rust/packages/create.html?highlight=custom%20package#create-a-custom-package) to instantiate the policy function calls only once.
+
+### Preliminary benchmark
+
+The following results are based on the [Kubewarden k6 load test](https://github.com/kubewarden/load-testing/tree/k6) using the [psp-apparmor policy](https://github.com/kubewarden/apparmor-psp-policy)
+
+|               |                          | http_req_duration |
+| ------------- | ------------------------ | ----------------- |
+| single policy | psp-apparmor             | avg=2.02ms        |
+| group policy  | `psp-apparmor() && true` | avg=2.46ms        |
+
+Note that no optimizations were made to the Rhai engine in this POC implementation.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- The implementation of the policy groups will add complexity to the Kubewarden controller and the policy server.
+- The expression language used to evaluate the policy groups may not be familiar to all users.
+- In the worst case, the policy group evaluation time could be the sum of the evaluation time of all the policies in the group.
+  This is even worse if the policy group contains context-aware policies that introduce additional overhead.
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+An alternative to the policy groups could be to create a tool that takes several policies and merges them into a new WASN module.
+The new WASM module would embed the original ones and would have a `main` function that takes care of evaluating the expression given by the user.
+
+This approach has the following drawbacks:
+
+- Complexity to implement this solution
+- The user would require an extra compilation step to create the new WASM module
+- The size of the new WASM module would be significantly larger if several policies are embedded
+- At runtime, the de-duplication optmization of the policies would be lost, increasing the memory footprint
+
+# Unresolved questions
```

Comment:
Are policy groups compatible with raw policies? (that return a json)
Do we foresee supporting policies in policy groups that return valid CEL, to be evaluated in `expression`?
It seems that's the case, and this implementation API is forward compatible for that, hence maybe worth a mention.

Reply:
policy group are fully compatbile with raw policies out of the box, I will add a paragraph to clarify this